### ISSUE TITLE
Batched MAGMA calls illegally read CUDA memory

### ISSUE THREAD

**Issue by mruberry** [pytorch/pytorch#26996]
## 🐛 Bug

(Some?) Batched MAGMA calls illegally read CUDA memory.

These illegal reads are often "silent" and harmless. If, however, they access unallocated device memory they will cause the program's future CUDA calls to fail. 

## To Reproduce

See #26789. Or once that PR goes in just re-enable test_cholesky_batched_many_batches.

This can also be reproduced by calling magma_dpotrf_batched directly with a tensor allocated by cudaMalloc. Run under cuda-memcheck to report all illegal memory accesses, including "silent" ones.

## Additional context

This issue was discovered in #26789 and diagnosed by me and @ngimel. We are following-up with @vishwakftw. 

A workaround may be to pad tensor inputs to batched MAGMA calls. This requires copying the original tensor into a buffer larger than it needs. How much extra space is needed requires further investigation. 

cc @ezyang @gchanan @zou3519 @vincentqb @vishwakftw @jianyuh @SsnL

**Comment by richardk53** [comment]
Any updates on this issue? 
This issue has also been reported [here](https://discuss.pytorch.org/t/cuda-illegal-memory-access-when-using-batched-torch-cholesky/51624/8).

**Comment by vishwakftw** [comment]
I haven’t gotten to this issue due to prior commitments.

**Comment by richardk53** [comment]
I think this issue should be assigned a high priority label as it causes the program to crash. 
It is possibly related to #36921 and others?

**Comment by vishwakftw** [comment]
Yes it is related. The errors that you are seeing are propagated from MAGMA, and not PyTorch. The implementation is simply a wrapper around the batches routines in MAGMA.

**Comment by y0ast** [comment]
I just hit this today and was looking at the magma issues, it doesn't seem to be reported there yet, is that correct?

I also found this to be significantly worse in pytorch 1.6.0 vs 1.5.1, is that possible? I could manage many more calls to `torch.cholesky` until I hit the memory error before the update.

It's a fairly painful bug that only shows up under odd circumstances (decreasing batch size which leads to increasing number of steps/epoch and therefore `torch.cholesky` calls for example).

**Comment by mruberry** [comment]
Thanks for this update @y0ast. I've bumped this issue's priority.

**Comment by Balandat** [comment]
Chiming in here, this is happening quite frequently in our runs (for `choleksy` and it seems for `symeig` as well) and is a major issue. @vishwakftw is there a MAGMA issue tracking this?

**Comment by gchanan** [comment]
have we reported this to the MAGMA team?

**Comment by ngimel** [comment]
We did not report this particular error to magma (as it happens randomly), but our other report did not get a lot of attention https://bitbucket.org/icl/magma/issues/19/dgesv_batched-dgetrs_batched-fails-for

**Comment by richardk53** [comment]
> We did not report this particular error to magma (as it happens randomly), but our other report did not get a lot of attention https://bitbucket.org/icl/magma/issues/19/dgesv_batched-dgetrs_batched-fails-for

it does not happen randomly though. 
In my setups, these errors were thrown for certain batch sizes.

**Comment by mruberry** [comment]
Can you elaborate on which functions and which batch sizes you're seeing segfaults, @richardk53? It would be very helpful in verifying our thinking.

**Comment by richardk53** [comment]
In my case it was a neural network with particle filter inference. 
The matrices for which the Magma calls sometimes failed hat a leading Particle (P) and Batch (B) dimension. 
For example, certain neural network architectures architectures always failed with B=32 and P=32, but never failed with B=33 and P=32. 
But this was neural network architecture dependent (as it leads to a different memory usage). 
Also, it is *not* the case that it only happens for exponentials of 2 (e.g. B=2^X and/or P=2^Y). 

The neural network architectures I had trouble with are fairly complicated; it is a time-series model and the inference algorithm is a rao-blackwellized particle filter, involving cholesky decomposition and cholesky-based matrix inversions. 
It should be (I hope) possible to get a much simpler minimal example...

**Comment by ngimel** [comment]
@richardk53 if you have pytorch 1.7 or source build, can you try your workload with PYTORCH_NO_CUDA_MEMORY_CACHING=1 and report failing sizes? This should disable caching allocator and make magma failures more frequent and predictable. Thank you!

**Comment by mruberry** [comment]
Closing now that #50957 is in.
